### PR TITLE
Wallet history

### PR DESCRIPTION
**Pull request by chris-belcher**
Implementation of #200 

You need to be configured to use Bitcoin Core (-txindex=1 not required), run it with:

`python wallet-tool.py wallet.json history`

Prints a summary for every transaction. If you have numpy/scipy installed it also calculates the effective interest rate you achieved as if yield-generators were a savings account. For my own yield generator wallet the figure is 0.67% per annum

It checks if the amount of bitcoins in the wallet is the same as the resulting amount when added up as iterating through all transactions. It also checks if the number of UTXOs in the wallet match the number found by iterating through transactions. Both these checks fail for me, and I can't figure out why.

You can create a csv file for opening with spreadsheet software too:

`python wallet-tool.py --csv wallet.json history >> history.csv`

### CODE REVIEW

**Comment by AdamISZ** [comment]
Appears to be well-isolated, can't see any impact here, so utACK.

**Comment by AdamISZ** [comment]
Quick record of test result as discussed on IRC:

Csv data was created OK.

>It checks if the amount of bitcoins in the wallet is the same as the resulting amount when added up as iterating through all transactions. It also checks if the number of UTXOs in the wallet match the number found by iterating through transactions. Both these checks fail for me, and I can't figure out why.

Same here, these checks failed.
Additional vague anecdotal: the number of "cj withdraw" looks too high (I see 21, expect < 15), the number of "deposit" also (see 14 but probably around 10). Sorry for vagueness, memory not good enough.

`getblockheader` was not found running Bitcoin Core 0.11.2, had to use `getblock` instead. So presumably either account for this (including the intersection with pruning) or mention somewhere that 0.12 is required (unfortunately not yet recorded [here](https://bitcoin.org/en/developer-reference#rpcs)).

**Comment by chris-belcher** [comment]
Edited to fall back to `getblock` if `getblockheader` doesn't exist.

I checked my cj withdraw transactions very carefully to find some error but everything seemed to work. Although it sounds like waxwing's problem is too many withdraws rather than incorrectly the calculated balance.

**Comment by chris-belcher** [comment]
grubles from IRC tried this (and got 3.2% as an effective interest rate) but didn't get any of the errors. He only deposited coins into the wallet and didnt withdraw, which suggests again it's a problem with `cj withdrawal` somehow.

**Comment by chris-belcher** [comment]
An idea for finding the bug, for me the UTXO count from history is higher than the wallet UTXO count. So history believes there are extra UTXOs which dont actually exist. So if I compare the two UTXO sets and find their difference it should give me some clues to what went wrong.

**Comment by chris-belcher** [comment]
Used the above idea to find and fix the bug.
Turns out some transactions are in walletname="", in my case I had exported a privatekey and spent the UTXO entirely to some other address. This didnt appear with the label "joinmarket-wallet-whatever".

Seems to work now. Please, everybody try this.

**Comment by AdamISZ** [comment]
Works correctly, feel free to merge.

File: wallet-tool.py [hunk 1]
```diff
@@ -269,3 +282,200 @@ def cus_print(s):
...
+        def f(r, deposits, deposit_times, now, final_balance):
+            return np.sum(np.exp((now - deposit_times) / 60.0 / 60 / 24 /
+                365)**r * deposits) - final_balance
+        r = brentq(f, a=0.1, b=-0.1, args=(deposits, deposit_times, now,
+            balance))
+        print('continuously compounded equivalent annual interest rate = ' +
+            str(r * 100) + ' %')
+        print('(as if yield generator was a bank account)')
+    except ImportError:
+        print('numpy/scipy not installed to unable to calculate effective ' +
```
Comment (chris-belcher) [inline wallet-tool.py]: spelling mistake here

**Comment by chris-belcher** [comment]
note to self: must add history to the --help documentation

**Comment by chris-belcher** [comment]
Done everything, ready to merge except for two extra commits that have appeared in this PR. Somehow their hash must've changed because of me doing git rebase

I have half a mind to merge anyway.

**Pull request merged**